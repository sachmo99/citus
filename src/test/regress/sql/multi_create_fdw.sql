
SET citus.next_shard_id TO 390000;


-- ===================================================================
-- get ready for the foreign data wrapper tests
-- ===================================================================

-- create fake fdw for use in tests
CREATE FUNCTION fake_fdw_handler()
RETURNS fdw_handler
AS 'citus'
LANGUAGE C STRICT;

set citus.enable_ddl_propagation to off;
CREATE FOREIGN DATA WRAPPER fake_fdw HANDLER fake_fdw_handler;
CREATE SERVER fake_fdw_server FOREIGN DATA WRAPPER fake_fdw;
set citus.enable_ddl_propagation to on;

-- test propagating foreign server creation
CREATE EXTENSION postgres_fdw;
CREATE SERVER foreign_server TYPE 'test_type' VERSION 'v1'
        FOREIGN DATA WRAPPER postgres_fdw
        OPTIONS (host 'testhost', port '5432', dbname 'testdb');

SELECT COUNT(*)=1 FROM pg_foreign_server WHERE srvname = 'foreign_server';
\c - - - :worker_1_port
-- verify that the server is created on the worker
SELECT COUNT(*)=1 FROM pg_foreign_server WHERE srvname = 'foreign_server';
\c - - - :master_port
ALTER SERVER foreign_server OPTIONS (ADD passfile 'to_be_dropped');
ALTER SERVER foreign_server OPTIONS (SET host 'localhost');
ALTER SERVER foreign_server OPTIONS (DROP port, DROP dbname);
ALTER SERVER foreign_server OPTIONS (ADD port :'master_port', dbname 'regression', DROP passfile);
ALTER SERVER foreign_server RENAME TO foreign_server_1;

-- test alter owner
SELECT srvowner FROM pg_foreign_server WHERE srvname = 'foreign_server_1';
ALTER SERVER foreign_server_1 OWNER TO pg_monitor;
\c - - - :worker_1_port
-- verify that the server is renamed on the worker
SELECT srvoptions FROM pg_foreign_server WHERE srvname = 'foreign_server_1';
-- verify the owner is changed
SELECT srvowner FROM pg_foreign_server WHERE srvname = 'foreign_server_1';
ALTER SERVER foreign_server_1 OWNER TO postgres;
\c - - - :master_port

-- verify the owner is changed on the worker
SELECT srvowner FROM pg_foreign_server WHERE srvname = 'foreign_server_1';
DROP SERVER IF EXISTS foreign_server_1 CASCADE;
\c - - - :worker_1_port
-- verify that the server is dropped on the worker
SELECT COUNT(*)=0 FROM pg_foreign_server WHERE srvname = 'foreign_server_1';
\c - - - :master_port
