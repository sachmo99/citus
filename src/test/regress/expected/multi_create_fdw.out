SET citus.next_shard_id TO 390000;
-- ===================================================================
-- get ready for the foreign data wrapper tests
-- ===================================================================
-- create fake fdw for use in tests
CREATE FUNCTION fake_fdw_handler()
RETURNS fdw_handler
AS 'citus'
LANGUAGE C STRICT;
CREATE FOREIGN DATA WRAPPER fake_fdw HANDLER fake_fdw_handler;
CREATE SERVER fake_fdw_server FOREIGN DATA WRAPPER fake_fdw;
-- test propagating foreign server creation
CREATE EXTENSION postgres_fdw;
CREATE SERVER foreign_server TYPE 'test_type' VERSION 'v1'
        FOREIGN DATA WRAPPER postgres_fdw
        OPTIONS (host 'testhost', port '5432', dbname 'testdb');
SELECT COUNT(*)=1 FROM pg_foreign_server WHERE srvname = 'foreign_server';
 ?column?
---------------------------------------------------------------------
 t
(1 row)

\c - - - :worker_1_port
-- verify that the server is created on the worker
SELECT COUNT(*)=1 FROM pg_foreign_server WHERE srvname = 'foreign_server';
 ?column?
---------------------------------------------------------------------
 t
(1 row)

\c - - - :master_port
ALTER SERVER foreign_server OPTIONS (ADD passfile 'to_be_dropped');
ALTER SERVER foreign_server OPTIONS (SET host 'localhost');
ALTER SERVER foreign_server OPTIONS (DROP port, DROP dbname);
ALTER SERVER foreign_server OPTIONS (ADD port :'master_port', dbname 'regression', DROP passfile);
ALTER SERVER foreign_server RENAME TO "foreign'server_1!";
-- test alter owner
SELECT srvowner FROM pg_foreign_server WHERE srvname = 'foreign''server_1!';
 srvowner
---------------------------------------------------------------------
       10
(1 row)

ALTER SERVER "foreign'server_1!" OWNER TO pg_monitor;
\c - - - :worker_1_port
-- verify that the server is renamed on the worker
SELECT srvoptions FROM pg_foreign_server WHERE srvname = 'foreign''server_1!';
                  srvoptions
---------------------------------------------------------------------
 {host=localhost,port=57636,dbname=regression}
(1 row)

-- verify the owner is changed
SELECT srvowner FROM pg_foreign_server WHERE srvname = 'foreign''server_1!';
 srvowner
---------------------------------------------------------------------
     3373
(1 row)

-- this doesn't error out for now, since we don't have object metadata on the worker
ALTER SERVER "foreign'server_1!" OWNER TO postgres;
\c - - - :master_port
-- verify the owner is changed on the worker
SELECT srvowner FROM pg_foreign_server WHERE srvname = 'foreign''server_1!';
 srvowner
---------------------------------------------------------------------
     3373
(1 row)

DROP SERVER IF EXISTS "foreign'server_1!" CASCADE;
\c - - - :worker_1_port
-- verify that the server is dropped on the worker
SELECT COUNT(*)=0 FROM pg_foreign_server WHERE srvname = 'foreign''server_1!';
 ?column?
---------------------------------------------------------------------
 t
(1 row)

\c - - - :master_port
